### PR TITLE
[testing] first attempt at documenting our testing story

### PR DESCRIPTION
**Pull request by yjbanov**

### CODE REVIEW

File: testing.md [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+      // aggregate numbers, such as "average frame build time".
+      TimelineSummary summary = new TimelineSummary.summarize(timeline);
+      summary.writeSummaryToFile('stocks_scroll_perf', pretty: true);
+      summary.writeTimelineToFile('stocks_scroll_perf', pretty: true);
+    });
+  });
+}
+```
+
+### Run test
```
Comment (qchong) [inline testing.md]: suggest: "Running your tests".

File: testing.md [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+title: Testing
+permalink: /testing/
+---
+
+* TOC
+{:toc}
+
+## Introduction
+
+The more features your application has, the harder it is to test it manually. A
```
Comment (qchong) [inline testing.md]: I noticed you use the term "application" throughout this doc and in your main title. It's acceptable to shorten this to "app" which is the more common term for mobile (developers know that Flutter apps are for mobile devices).

File: testing.md (hunk 1 above)
Comment (sethladd) [inline testing.md]: +1

Using "...ing" is a good style for the headers of a doc like this.

File: testing.md [hunk 3]
```diff
@@ -0,0 +1,299 @@
+---
+layout: page
+title: Testing
```
Comment (sethladd) [inline testing.md]: For make better SEO, did we consider "Testing Flutter" or "Testing Flutter apps" ?

File: testing.md [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+---
+
+* TOC
+{:toc}
+
+## Introduction
+
+The more features your application has, the harder it is to test it manually. A
+good set of automated tests will help you make sure your application performs
+correctly when you publish it to users while retaining your feature and bug fix
```
Comment (qchong) [inline testing.md]: s/when/before/

You can drop "to users".

File: testing.md [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+| **Execution speed**  | Quick  | Slower | Slowest    |
+
+As a rule of thumb a well-tested application has a very high number of unit
+and widget tests, tracked by line coverage, and a good number of end-to-end
+tests covering all the important usage scenarios.
+
+
+## Unit testing
+
+**NOTE: `flutter test` currently only works on Linux**
```
Comment (sethladd) [inline testing.md]: Is there a bug for Mac, that we can link to?

File: testing.md [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+import 'package:test/test.dart';
+
+void main() {
+  test('this is a unit test', () {
+    expect(42, 42);
+  });
+}
+```
+
+Run `flutter test test/unit_test.dart`.
```
Comment (sethladd) [inline testing.md]: But the above test doesn't depend on flutter framework. Couldn't I have just run "dart test/unit_test.dart" ?

File: testing.md [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+{:toc}
+
+## Introduction
+
+The more features your application has, the harder it is to test it manually. A
+good set of automated tests will help you make sure your application performs
+correctly when you publish it to users while retaining your feature and bug fix
+velocity.
+
+There are many kinds of automated testing. Their definitions are fuzzy, so let's
```
Comment (qchong) [inline testing.md]: For brevity, you can drop the whole sentence beginning with "Their definitions ..." up to the colon, and instead say "These are summarized below:"

File: testing.md [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+A Flutter end-to-end test is also written using `package:test`. A full test is a
+pair - a test script and a Flutter application instrumented to receive commands
+from the test. Unlike unit and widget tests, end-to-end test code does not run
+in the same process as the application that's being tested. Instead, the tested
+application is launched on a _real device_ or in an _emulator_ (e.g. Android
+simulator or iOS Simulator). The test script runs on your computer. It connects
+to the application and issues commands to the application to perform various
+user actions. This is known as "driving" the application. Flutter provides tools
+and API, collectively referred to as _Flutter Driver_, to do just that.
```
Comment (sethladd) [inline testing.md]: "and an API"

missing 'an'

File: testing.md [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+The more features your application has, the harder it is to test it manually. A
+good set of automated tests will help you make sure your application performs
+correctly when you publish it to users while retaining your feature and bug fix
+velocity.
+
+There are many kinds of automated testing. Their definitions are fuzzy, so let's
+define them here for clarity:
+
+- A _unit test_ tests a single function, method or class. External dependencies
+  of the test subject are generally mocked out using, for example,
```
Comment (qchong) [inline testing.md]: Instead of "test subject", I'd use the term "unit under test".

File: testing.md [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+
+The more features your application has, the harder it is to test it manually. A
+good set of automated tests will help you make sure your application performs
+correctly when you publish it to users while retaining your feature and bug fix
+velocity.
+
+There are many kinds of automated testing. Their definitions are fuzzy, so let's
+define them here for clarity:
+
+- A _unit test_ tests a single function, method or class. External dependencies
```
Comment (qchong) [inline testing.md]: nit: Please add a comma after "method". That makes it easier to parse a list of comma-separate items in the same sentence.

File: testing.md [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+  of the test subject are generally mocked out using, for example,
+  [`package:mockito`](https://github.com/fibulwinter/dart-mockito/blob/master/README.md).
+  Unit tests generally do not read from/write to disk, render to screen and do
+  not receive user actions from outside the process running the test. The goal
+  of a unit-test is to verify the correctness of a piece of logic under a
```
Comment (qchong) [inline testing.md]: Drop the hyphen from "unit test".

File: testing.md [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+  variety of conditions.
+- A _widget test_ (in other UI frameworks referred to as _component test_) tests
+  a single widget. Testing a widget involves multiple classes and requires a
```
Comment (qchong) [inline testing.md]: Is it always true that testing a widget involves multiple classes? What if there's only one StatelessWidget class that you're interested in testing?

File: testing.md [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+  test environment that makes the widget feel like it's part of a real UI. For
```
Comment (qchong) [inline testing.md]: "makes the widget feel like it's part of a real UI" <-- I'd tighten this phrasing (software doesn't have "feelings" and it's not necessarily clear what a "real UI" is). 

How about saying instead: "test environment that allows the widget to run in an app-like context."

File: testing.md (hunk 12 above)
Comment (krisgiesing) [inline testing.md]: Widget tests (for the most part) use other parts of the framework in order to exercise the widget. For example, widgets instantiate elements, which instantiate render objects. The elements and render objects aren't mocked out in widget tests. Other parts of the system may be used to drive the lifecycle of the objects involved.

File: testing.md [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+  example, it should be able to receive and respond to user actions and events,
+  perform layout, and instantiate child widgets. A widget test is
+  therefore more comprehensive than a unit test. However, like a unit test, a
+  widget test's environment is replaced with an implementation much simpler than
+  a full-blown UI system. The goal of a widget test is to verify that it looks
```
Comment (qchong) [inline testing.md]: suggest: "verify that the widget's UI looks"

File: testing.md [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+  and interacts as expected.
+- An _end-to-end test_ tests a complete application or a large part of an
```
Comment (qchong) [inline testing.md]: nit: I think "Integration test" is the more common term, would suggest to use that instead.

File: testing.md (hunk 12 above)
Comment (krisgiesing) [inline testing.md]: Widget tests still generally don't read or write files, render to screen, or involve user actions from outside the test. From that perspective they're very similar to unit tests.

File: testing.md [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+  application. Generally an _end-to-end test_ runs on a real device or an OS
```
Comment (qchong) [inline testing.md]: nit: Please add a comma after "Generally".

File: testing.md (hunk 13 above)
Comment (krisgiesing) [inline testing.md]: Or "that provides the appropriate widget lifecycle context".

File: testing.md (hunk 15 above)
Comment (krisgiesing) [inline testing.md]: I think this phrasing is fine, personally. There isn't an absolute accepted terminology in this space AFAIK.

File: testing.md [hunk 17]
```diff
@@ -0,0 +1,299 @@
...
+  emulator. The test code does not run in the same process (or even the same
+  device) as the test subject. The goal of an end-to-end test is to verify that
```
Comment (qchong) [inline testing.md]: "(or even the same device)" <-- this is a bit confusing since in the previous sentence you defined an e2e test as one that runs on a "real device". 

If the Flutter e2e test runtime environment is dependent on the mobile device platform, can you please specify accordingly? 
* For Flutter apps on Android, do the e2e tests run on the same device as the Flutter app?
* For Flutter apps on iOS, does the e2e test run on either the iOS simulator or same device as the Flutter app?

This distinction matters because it helps developers understand if there are restrictions for running their e2e tests (and plan accordingly).

File: testing.md [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+  the application functions correctly as a whole, that all the widgets it is
+  composed of integrate with each other as expected (which is why another term
+  for these kinds of tests is _integration test_). Another important aspect
+  end-to-end tests verify is application performance.
```
Comment (qchong) [inline testing.md]: Nit: Suggest rephrasing this sentence to make this more active. e.g. "You can also use your end-to-end tests to verify your app's performance."

File: testing.md [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+different kinds of tests:
+
+|                      | Unit   | Widget | End-to-end |
+|----------------------|--------|--------|------------|
+| **Effectiveness**    | Low    | Higher | Highest    |
+| **Maintenance cost** | Low    | Higher | Highest    |
+| **Dependencies**     | Few    | More   | Lots       |
+| **Execution speed**  | Quick  | Slower | Slowest    |
+
+As a rule of thumb a well-tested application has a very high number of unit
```
Comment (qchong) [inline testing.md]: This is a very good recommendation. Suggest highlighting it by adding a bold **Tip:** prefix (before "As a rule...").

File: testing.md [hunk 20]
```diff
@@ -0,0 +1,299 @@
...
+and widget tests, tracked by line coverage, and a good number of end-to-end
```
Comment (qchong) [inline testing.md]: The term "line coverage" was unfamiliar to me until Yegor explained it. Suggest adding a link to a brief explainer (like this page: http://www.softwaretestingmentor.com/test-design-techniques/line-coverage/)

File: testing.md (hunk 3 above)
Comment (yjbanov) [inline testing.md]: Changed to "Testing Flutter apps"

File: testing.md [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+
+Run `flutter test test/unit_test.dart`.
+
+
+## Widget testing
+
+**NOTE: `flutter test` currently only works on Linux**
+
+A widget test is implemented just like a unit test. To let the test interact with
+the tested widget, Flutter provides a utility
```
Comment (qchong) [inline testing.md]: Suggest tightening this paragraph as follows:

"You implement a widget test in a similar way as a unit test. To perform an interaction with a widget in your test, use the `WidgetTester` utility that Flutter provides. For example, you can send tap and scroll gestures. You can also use `WidgetTester` to find child widgets in the widget tree, read text, and verify that the values of widget properties are correct."

File: testing.md (hunk 2 above)
Comment (yjbanov) [inline testing.md]: Done.

File: testing.md (hunk 4 above)
Comment (yjbanov) [inline testing.md]: Done.

File: testing.md (hunk 7 above)
Comment (yjbanov) [inline testing.md]: Done.

File: testing.md (hunk 10 above)
Comment (yjbanov) [inline testing.md]: Done.

File: testing.md (hunk 9 above)
Comment (yjbanov) [inline testing.md]: Done.

File: testing.md (hunk 11 above)
Comment (yjbanov) [inline testing.md]: Done.

File: testing.md (hunk 12 above)
Comment (yjbanov) [inline testing.md]: +1 to what @krisgiesing said

File: testing.md [hunk 22]
```diff
@@ -0,0 +1,299 @@
...
+* build the `--target` application and install it on the device
+* launch the application
+* run the `my_app/test_driver/user_list_scrolling_test.dart` test
+
+You might be wondering how the command finds the correct test file. The
+`flutter drive` command uses a convention to look for the test file in the same
+directory as the instrumented `--target` application that has the same file name
+but for the `_test` suffix in it.
+
+**NOTE: We currently only support Android devices and iOS Simulator. Support for
```
Comment (qchong) [inline testing.md]: Please move this note further above, right after where you first describe the Flutter driver.

**Comment by qchong** [comment]
@yjbanov Completed my first pass at editorial review. Overall, this information looks good!

File: testing.md (hunk 13 above)
Comment (yjbanov) [inline testing.md]: Went with Kris' version.

File: testing.md (hunk 14 above)
Comment (yjbanov) [inline testing.md]: Done

File: testing.md (hunk 16 above)
Comment (yjbanov) [inline testing.md]: Done.

File: testing.md (hunk 15 above)
Comment (yjbanov) [inline testing.md]: Going with "integration test". It has a [Wikipedia article](https://en.wikipedia.org/wiki/Integration_testing)! :)

File: testing.md (hunk 17 above)
Comment (yjbanov) [inline testing.md]: > "(or even the same device)" <-- this is a bit confusing since in the previous sentence you defined an e2e test as one that runs on a "real device".

I changed the wording. The point I'm trying to convey is: in an integration test, the app is isolated from the test code _somehow_, be it separate thread, process, or hardware. This is to avoid skewing the test results by mixing the test code and the tested code (i.e. minimize the [observer effect](https://en.wikipedia.org/wiki/Observer_effect_(physics))).

I avoided going into Flutter-specific solutions on purpose. The purpose of this section is to give a general definition of what an integration test is. Flutter specifics are discussed in the dedicated section below.

File: testing.md (hunk 18 above)
Comment (yjbanov) [inline testing.md]: Done.

File: testing.md (hunk 19 above)
Comment (yjbanov) [inline testing.md]: Done.

File: testing.md (hunk 20 above)
Comment (yjbanov) [inline testing.md]: Switched to Wikipedia-friendly "code coverage".

File: testing.md (hunk 5 above)
Comment (yjbanov) [inline testing.md]: Filed: https://github.com/flutter/flutter/issues/3179

File: testing.md (hunk 6 above)
Comment (yjbanov) [inline testing.md]: Refactored the whole paragraph

File: testing.md (hunk 8 above)
Comment (yjbanov) [inline testing.md]: That depends on whether "I" stands for "interface" or "interfaces" :) Some people write "APIs", but I don't like that for some reason.

@qchong

File: testing.md (hunk 1 above)
Comment (yjbanov) [inline testing.md]: Cool. I'll similarly rephrase the sibling headers.

**Comment by yjbanov** [comment]
Blocked by https://github.com/flutter/flutter/pull/3205 (the PR has to make it into alpha branch before this PR can be submitted)

**Comment by yjbanov** [comment]
@qchong, @sethladd can haz lgtm?

**Comment by qchong** [comment]
lgtm
